### PR TITLE
Allow to use page URL in pagination patterns

### PR DESCRIPTION
**Pull request by mosra**
Currently it was only possible to use page "save as" name or part of it for generating pagination links. That's not sufficient when page URLs differ a lot from actual filenames. With this patch it's possible to use the `{url}` placeholder in the `PAGINATION_PATTERNS` setting. For example, the paginated archives would be saved to:

    blog/index.html
    blog/2/index.html
    blog/3/index.html

while the actual URLs would be like this (with the help of Apache's `mod_rewrite`, for example):

    http://blog.my.site/
    http://blog.my.site/2/
    http://blog.my.site/3/

The configuration that corresponds to this is roughly the following:

```py
ARCHIVES_SAVE_AS = 'blog/index.html'
ARCHIVES_URL = 'http://blog.my.site/'
PAGINATION_PATTERNS = [
    (1, '/{url}', '{base_name}/index.html'),
    (2, '/{url}{number}/', '{base_name}/{number}/index.html')
]
```

**Note:** I added a test case for this, but didn't update any docs as the various placeholders are not mentioned anywhere in the documentation so I didn't know where to put that. Possibly related (regarding documentation): #1137 #1615

Thank you!

### CODE REVIEW

**Comment by mosra** [comment]
Okay, the Travis builds fails because of flake8 checks, but these errors are the same as with current master and I assume something got different on Travis and caused these:

```
pelican/tests/support.py:58:15: E741 ambiguous variable name 'l'
pelican/tools/pelican_quickstart.py:17:1: E722 do not use bare except'
pelican/tools/pelican_quickstart.py:117:52: E741 ambiguous variable name 'l'
pelican/tools/pelican_quickstart.py:183:13: E722 do not use bare except'
pelican/tools/pelican_themes.py:20:1: E722 do not use bare except'
```

**Comment by justinmayer** [comment]
Thanks, @mosra. I fixed the PyCodeStyle issues via 56a48347 and have re-triggered your PR's build, so everything is green here now. I will take a look at this PR as soon as I can.

**Comment by mosra** [comment]
Great, thank you!

**Comment by justinmayer** [comment]
My admittedly cursory scan of this pull request didn't yield any glaring problems. Any @getpelican/reviewers care to comment before this is merged?

File: pelican/writers.py [hunk 1]
```diff
@@ -140,11 +140,12 @@ def write_feed(self, elements, context, path=None, feed_type='atom',
                 complete_path, context=context, feed=feed)
         return feed
 
-    def write_file(self, name, template, context, relative_urls=False,
+    def write_file(self, name, url, template, context, relative_urls=False,
```
Comment (avaris) [inline pelican/writers.py]: This is quite a change and could possibly break some plugins. Did you do any checks? Somewhat *nicer* way would be making this parameter kwarg.

Comment (mosra) [reply pelican/writers.py]: Made `url` into an optional parameter, defaulting to `None`. Is that okay? :)

Comment (avaris) [reply pelican/writers.py]: Yeah, should be fine now. Thanks.

**Comment by justinmayer** [comment]
Excellent description of the problem to be solved, references to related issues, documentation, and test coverage. If only all pull requests looked like this one! 😁 

Bravo, @mosra. Many thanks for the contribution. (And thanks @avaris for reviewing!)

**Pull request merged**